# Hello TuiML

**Your first 10 minutes with TuiML — agent-first, end-to-end.**

The homepage promises:

> *"Ask your agent to train a model, tune it, compare it to the last run, or find an algorithm that fits your data. It just does it. No code. No guesswork. No forgotten context."*

This tour walks every clause of that sentence — *find* → *train* → *tune* → *compare* → *deploy* — using a real medical dataset and the same tools an agent calls over MCP.

Each section is mapped to one of TuiML's six homepage pillars so you can see the promise → tool wiring directly. Plus integrations for **LangChain** and **Pydantic-AI** at the end.

**Total time:** ~10 min reading, ~2 min execution. Runs end-to-end with no API key.

---

## 0. Setup — simulate or go live

By default this notebook **simulates** the agent by calling tools through `execute_tool` directly. Set `MODE = "live"` and add your Anthropic API key to drive everything with a real Claude model in the loop. The tool calls are identical either way — that's the point.

> 🟦 **Pillar 1 · 100% local & private.** Even in `live` mode, your data never leaves the machine — only the prompts and tool *names* go to the LLM, never the dataset rows.

In [1]:
MODE = "simulate"   # or "live" — needs ANTHROPIC_API_KEY in env
DATASET = "breast-cancer"
TARGET = "Class"

from tuiml.agent import execute_tool, get_tools_for_llm
print(f"Running in {MODE} mode | dataset = {DATASET} | {len(get_tools_for_llm())} tools available")

Running in simulate mode | dataset = breast-cancer | 27 tools available


---
## 1. Find an algorithm that fits the data  *(1.5 min)*

> 🟦 **Pillar 2 · Algorithm Discovery.** Agents don't memorise class names. They profile the data, then search the catalog by task and constraints.

> **User:** *"I have a dataset called `breast-cancer`. Look at it and recommend 2-3 classifiers that should work well."*

**Tool call 1.1:** `tuiml_data_profile`

In [2]:
profile = execute_tool("tuiml_data_profile", data=DATASET, target=TARGET)
print(f"Shape:   {profile['shape']}")
print(f"Features: {profile['n_features']}")
print(f"Missing: {profile['missing_values']}")
print(f"First 3 features: {profile['feature_names'][:3]}")

Shape:   [286, 9]
Features: 9
Missing: {'node-caps': 8, 'breast-quad': 1}
First 3 features: ['age', 'menopause', 'tumor-size']


**Agent's reasoning:** *"286 samples, 9 categorical-encoded features, some missing values. Small dataset — favours simple, regularised models. Let me see what's in the catalog."*

**Tool call 1.2:** `tuiml_search`

In [3]:
candidates = []
for query in ["random forest", "naive bayes", "logistic"]:
    hits = execute_tool("tuiml_search", query=query, category="algorithm")
    for h in hits["results"][:1]:
        name = h["name"].replace("tuiml_algorithm_", "")
        candidates.append(name)
        print(f"  {name:<32} {h['description'][:50]}")
print(f"\nAgent's shortlist: {candidates}")

  RandomForestClassifier           Random Forest classifier - ensemble of random tree
  NaiveBayesClassifier             Naive Bayes classifier using **pluggable probabili
  LogisticModelTreeClassifier      Logistic Model Trees classifier.

Agent's shortlist: ['RandomForestClassifier', 'NaiveBayesClassifier', 'LogisticModelTreeClassifier']


---
## 2. Train a baseline  *(1 min)*

> 🟦 **Pillar 3 · Train · Tune · Compare.** First piece: train. One tool call, one model, full metrics returned.

> **User:** *"Train the random forest with 5-fold CV. Save the result as my baseline."*

**Tool call 2.1:** `tuiml_train`

In [4]:
baseline = execute_tool(
    "tuiml_train",
    algorithm="RandomForestClassifier",
    data=DATASET,
    target=TARGET,
    cv=5,
)
baseline_acc = baseline["metrics"]["cv_accuracy_score_mean"]
baseline_id = baseline["model_id"]
print(f"Baseline: {baseline['model_class']}")
print(f"  CV accuracy: {baseline_acc:.4f}")
print(f"  Model ID:   {baseline_id}")

Baseline: RandomForestClassifier
  CV accuracy: 0.7167
  Model ID:   bda0d6d460fe


---
## 3. Tune it  *(2 min)*

> 🟦 **Pillar 3 · Train · Tune · Compare.** The middle piece. Agents construct search spaces from the algorithm's typed parameter schema — no hallucinated hyperparameter names.

> **User:** *"Now tune the random forest. Pick the search space yourself based on what the algorithm supports."*

**Tool call 3.1:** `tuiml_describe` (read the schema first)

In [5]:
schema = execute_tool("tuiml_describe", name="RandomForestClassifier")
print("Tunable parameters (per the schema the agent reads):")
for name in ["n_estimators", "max_depth", "max_features"]:
    p = schema["parameters"][name]
    print(f"  {name:<14} type={p['type']}  default={p.get('default')}")

Tunable parameters (per the schema the agent reads):
  n_estimators   type=integer  default=100
  max_depth      type=integer  default=None
  max_features   type=['string', 'integer', 'number']  default=sqrt


**Agent's reasoning:** *"Schema says `n_estimators` is integer, `max_depth` integer or null. I'll grid over a sensible range."*

**Tool call 3.2:** `tuiml_tune`

In [6]:
tuned = execute_tool(
    "tuiml_tune",
    algorithm="RandomForestClassifier",
    data=DATASET,
    target=TARGET,
    method="grid",
    param_grid={
        "n_estimators": [50, 100, 200],
        "max_depth": [3, 5, None],
    },
    cv=5,
)
tuned_acc = tuned["best_score"]
tuned_id = tuned["model_id"]
print(f"Best params:   {tuned['best_params']}")
print(f"Tuned score:   {tuned_acc:.4f}  (baseline was {baseline_acc:.4f})")
print(f"Δ:              {tuned_acc - baseline_acc:+.4f}")

Best params:   {'n_estimators': 50, 'max_depth': 5}
Tuned score:   0.7590  (baseline was 0.7167)
Δ:              +0.0423


---
## 4. Compare it to the last run  *(1.5 min)*

> 🟦 **Pillar 4 · Persistent Experiments.** Models live on disk with their metadata. The agent loads yesterday's baseline, runs a head-to-head against today's tuned candidate, and picks the winner with a statistical test.

> **User:** *"Compare the tuned random forest to the baseline AND to a naive Bayes alternative. Run it under the same CV so it's apples-to-apples, and tell me which to ship."*

**Tool call 4.1:** `tuiml_experiment` (head-to-head with statistical comparison)

In [7]:
exp = execute_tool(
    "tuiml_experiment",
    algorithms=[
        "RandomForestClassifier",
        "NaiveBayesClassifier",
    ],
    data=DATASET,
    target=TARGET,
    cv=10,
)
print(exp["summary"])

Experiment: Experiment
Validation: cross_validation
Metric: accuracy_score

Dataset: breast-cancer
--------------------------------------------------
  RandomForestClassifier: 0.7168 ± 0.0773
  NaiveBayesClassifier: 0.7241 ± 0.0835



In [8]:
# Agent's deterministic decision rule — same code an agent would emit
scores = {
    name: m["accuracy_score"]["mean"]
    for name, m in exp["results"][DATASET].items()
}
winner = max(scores, key=scores.get)
print(f"Agent's decision: ship {winner} (acc={scores[winner]:.4f})")
print(f"Beats baseline by: {scores[winner] - baseline_acc:+.4f}")

Agent's decision: ship NaiveBayesClassifier (acc=0.7241)
Beats baseline by: +0.0074


---
## 5. Serve it  *(1 min)*

> 🟦 **Pillar 5 · One-Call Serving.** The model goes live as a REST endpoint with one tool call. No FastAPI scaffolding, no Docker file, no `if __name__ == '__main__'`.

> **User:** *"Save the winning model and serve it on localhost."*

**Tool call 5.1:** `tuiml_save_model`

In [9]:
saved = execute_tool(
    "tuiml_save_model",
    model_id=tuned_id if winner == "RandomForestClassifier" else baseline_id,
    destination="./tour_model.joblib",
)
print(f"Saved: {saved.get('destination')}")

Saved: /Users/nileshverma/Documents/GitHub/tuiml-project/tuiml-website/tutorials/llm_friendly/tour_model.joblib


**Tool call 5.2:** `tuiml_serve_model` *(commented out — uncomment in your own session)*

In [10]:
# served = execute_tool("tuiml_serve_model", model_path="./tour_model.joblib", port=8765)
# print(f"Endpoint: {served['url']}")
# Then from any client:
#   curl -X POST http://localhost:8765/predict \
#        -H 'Content-Type: application/json' \
#        -d '{"features": [[5,3,2,1,1,3,1,2,1]]}'
print("Skipped: would start a REST server on :8765.")
print("Stop with: execute_tool('tuiml_stop_server', port=8765)")

Skipped: would start a REST server on :8765.
Stop with: execute_tool('tuiml_stop_server', port=8765)


---
## 6. Drop into LangChain  *(1 min)*

> 🟦 **Pillar 6 · 200+ Algorithms · agents can call them.** TuiML's tools are framework-agnostic — wrap them once for LangChain and you have an MCP-style ML toolkit inside any LangChain agent.

The pattern: TuiML emits OpenAI-spec tool definitions; LangChain wraps them as `StructuredTool` instances. ~10 lines.

In [11]:
# Conceptual snippet — uncomment after `pip install langchain langchain-anthropic`
#
# from langchain_core.tools import StructuredTool
# from langchain_anthropic import ChatAnthropic
# from tuiml.agent import get_tools_for_llm, execute_tool
#
# def make_lc_tool(spec):
#     return StructuredTool.from_function(
#         name=spec["name"],
#         description=spec["description"],
#         func=lambda **kw: execute_tool(spec["name"], **kw),
#     )
#
# lc_tools = [make_lc_tool(t) for t in get_tools_for_llm()]
# llm = ChatAnthropic(model="claude-sonnet-4-6").bind_tools(lc_tools)
# response = llm.invoke("Train a random forest on breast-cancer with 5-fold CV.")
# print(response)
print("LangChain pattern shown above. Skipped to keep this notebook key-free.")

LangChain pattern shown above. Skipped to keep this notebook key-free.


---
## 7. Drop into Pydantic-AI  *(1 min)*

TuiML ships a pre-built Pydantic-AI agent with all 200+ tools and a SKILL prompt already loaded. One import.

In [12]:
# Uncomment after `pip install pydantic-ai`
#
# from tuiml.agent import Agent
# agent = Agent(model="anthropic:claude-sonnet-4-6")
# result = agent.run_sync("Profile breast-cancer, then train and tune the best classifier.")
# print(result.output)
print("Pydantic-AI agent: tuiml.agent.Agent is preloaded with every TuiML tool + SKILL.md prompt.")
print("CrewAI / AutoGen integrations follow the same pattern as LangChain — wrap get_tools_for_llm().")

Pydantic-AI agent: tuiml.agent.Agent is preloaded with every TuiML tool + SKILL.md prompt.
CrewAI / AutoGen integrations follow the same pattern as LangChain — wrap get_tools_for_llm().


---
## What you just saw

Every section maps back to one homepage promise:

| Section | Tool calls | Pillar |
|---|---|---|
| 1. Find an algorithm | `tuiml_data_profile`, `tuiml_search` | Algorithm Discovery |
| 2. Train a baseline | `tuiml_train` | Train · Tune · Compare |
| 3. Tune it | `tuiml_describe`, `tuiml_tune` | Train · Tune · Compare |
| 4. Compare runs | `tuiml_experiment` | Persistent Experiments |
| 5. Serve it | `tuiml_save_model`, `tuiml_serve_model` | One-Call Serving |
| 6 + 7. Integrations | `get_tools_for_llm`, `Agent` | 200+ Algorithms, framework-ready |

**Total: 7 tool calls, no glue code, one CSV's worth of state.** That's the homepage promise made concrete.

### Try it for real

1. Set `MODE = "live"` in cell 0 and add an `ANTHROPIC_API_KEY` to your environment.
2. Replace `execute_tool(...)` calls with `client.messages.create(...)` + the tool list — the agent will pick its own queries.
3. Or skip the SDK entirely: connect Claude Desktop via [MCP](/tutorials/llm_friendly/02_mcp_server.ipynb) and ask it the user prompts above directly.

### Where to go next

- **[Tools an agent can call](/tutorials/llm_friendly/01_llm_tools.ipynb)** — full provider conversion details (Claude, OpenAI, Gemini, Bedrock).
- **[Build an agentic workflow](/tutorials/llm_friendly/03_agentic_workflows.ipynb)** — autonomous loops with retries and longer task horizons.
- **[Case Study: Diabetes Prediction](/tutorials/case_studies/01_diabetes_prediction.ipynb)** — the same pattern on a richer dataset, end-to-end.

In [13]:
import os
if os.path.exists("tour_model.joblib"):
    os.remove("tour_model.joblib")